### PR TITLE
Add Global scope validator api

### PR DESCRIPTION
**Pull request by piraveena**
### Proposed changes in this pull request

- Resolves https://github.com/wso2/product-is/issues/8248

- This introduces a new interface to engage scope validators at the server/tenant level. Therefore all the validators implement this interface should Engage after application-level validators at response type handler level or grant level.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    /**
+     * Validates scopes in the token validation request and manipulate the permitted scopes within the request.
+     *
+     * @param accessTokenDO AccessTokenDO.
+     * @return True if the user has enough permission to access the resource with requested scopes or no scopes are
+     * requested, otherwise false.
+     * @throws IdentityOAuth2Exception
+     */
+    boolean validateScope(AccessTokenDO accessTokenDO) throws IdentityOAuth2Exception;
```
Comment (malithie) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java:69]: Let's not have this API at all

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ * This is a global level interface for scope validation. This needs to be engaged after application level validators.
+ */
+public interface ScopeValidator {
+
+    /**
+     * Checks whether the validator can be engaged.
+     *
+     * @return True if it can handle, otherwise false.
+     */
+    boolean canHandle();
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java:36]: How do we intend this API to be used?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+    /**
+     * Validates scopes in the token request and manipulate the permitted scopes within the request.
+     *
+     * @param tokReqMsgCtx OAuthTokenReqMessageContext.
+     * @return True if the user has enough permission to generate tokens with requested scopes or
+     * no scopes are requested, otherwise false.
+     * @throws IdentityOAuth2Exception
+     */
+    boolean validateScope(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception,
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java:57]: Do we need to allow UserStoreException to be thrown from the interface?

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java (hunk 2 above)
Comment (piraveena) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java:36]: can read the properties from config file or can have its own implementation specific to the validator implementations.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    /**
+     * Validates scopes in the authorization  request and manipulate the permitted scopes within the request.
+     *
+     * @param authzReqMessageContext Authorization request.
+     * @return True if the user has enough permission to generate tokens or authorization codes with requested
+     * scopes or no scopes are requested, otherwise false.
+     * @throws IdentityOAuth2Exception
+     */
+    boolean validateScope(OAuthAuthzReqMessageContext authzReqMessageContext) throws UserStoreException,
```
Comment (malithie) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java:46]: Should throw only 'IdentityOAuth2Exception'

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    /**
+     * Validates scopes in the authorization  request and manipulate the permitted scopes within the request.
+     *
+     * @param authzReqMessageContext Authorization request.
+     * @return True if the user has enough permission to generate tokens or authorization codes with requested
+     * scopes or no scopes are requested, otherwise false.
+     * @throws IdentityOAuth2Exception
+     */
+    boolean validateScope(OAuthAuthzReqMessageContext authzReqMessageContext) throws IdentityOAuth2Exception;
```
Comment (malithie) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java:46]: One more thing, shall we explain in the method description in which flow this API can be utilized. Applies to other overloaded validateScope apis as well

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+     *
+     * @param tokReqMsgCtx OAuthTokenReqMessageContext.
+     * @return True if the user has enough permission to generate tokens with requested scopes or
+     * no scopes are requested, otherwise false.
+     * @throws IdentityOAuth2Exception
+     */
+    boolean validateScope(OAuthTokenReqMessageContext tokReqMsgCtx) throws IdentityOAuth2Exception;
+
+    /**
+     * Validates scopes in the token request and manipulate the permitted scopes within the request.
```
Comment (malithie) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java:59]: Shall we say 'token validation request' here

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+    /**
+     * Checks whether the validator can be engaged.
+     *
+     * @return True if it can handle, otherwise false.
+     */
+    boolean canHandle();
+
+    /**
+     * Validates scopes in the authorization  request and manipulate the permitted scopes within the request. Engage
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/validators/scope/ScopeValidator.java:39]:
```suggestion
     * Validates scopes in the authorization request and manipulate the permitted scopes within the request. Engage
```

**Pull request merged**